### PR TITLE
 New API and fixes for group data.

### PR DESCRIPTION
**Pull request by AndreiKingsley**

### CODE REVIEW

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/GroupByPlotContext.kt [hunk 1]
```diff
@@ -22,10 +22,10 @@ import org.jetbrains.kotlinx.kandy.ir.feature.PlotFeature
...
  * @property plotFeatures A mutable map to store and manage features associated with the plot.
  */
-public class GroupByPlotContext<T>(
-    private val groupBy: GroupBy<T, T>
-) : LayerPlotContext(), ColumnsContainer<T> by groupBy.concat() {
-    override val _plotContext: PlotContext = this
+public class GroupByPlotContext<T, G>(
+    groupBy: GroupBy<T, G>
+) : GroupedLayerPlotContext<T, G>(groupBy), ColumnsContainer<G> by groupBy.concat() {
+    //override val _plotContext: PlotContext = this
```
Comment (devcrocod) [inline kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/GroupByPlotContext.kt:28]: remove

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/DatasetHandler.kt [hunk 2]
```diff
@@ -32,7 +32,7 @@ import org.jetbrains.kotlinx.kandy.ir.data.TableData
  */
 public class DatasetHandler(
     public val initialDataset: TableData,
-    private val columnAsRefOnly: Boolean = false,
+    //private val columnAsRefOnly: Boolean = false,
```
Comment (devcrocod) [inline kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/DatasetHandler.kt:35]: remove

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/DatasetHandler.kt [hunk 3]
```diff
@@ -109,7 +109,7 @@ public class DatasetHandler(
      * @return The name of the added column.
      */
     public fun addColumn(column: ColumnReference<*>): String {
-        return if (column is DataColumn<*> && !columnAsRefOnly) {
+        return if (column is DataColumn<*>/* && !columnAsRefOnly*/) {
```
Comment (devcrocod) [inline kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/DatasetHandler.kt:112]: remove

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/addDataset.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+internal fun LayerCollectorContext.addDataset(df: DataFrame<*>): Int {
+    val datasetHandlers = plotContext.datasetHandlers
+    datasetHandlers.add(DatasetHandler(NamedData(df)))
+    return datasetHandlers.indices.last
+}
+
+internal fun LayerCollectorContext.addDataset(gb: GroupBy<*, *>, initialBuffer: DataFrame<*>): Int {
+    val datasetHandlers = plotContext.datasetHandlers
+    datasetHandlers.add(DatasetHandler(GroupedData(gb), initialBuffer))
+    return datasetHandlers.indices.last
```
Comment (devcrocod) [inline kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/addDataset.kt:17]: lastIndex?

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/ir/data/GroupedData.kt [hunk 5]
```diff
@@ -19,26 +19,28 @@ import org.jetbrains.kotlinx.dataframe.api.groupBy
...
  */
 public data class GroupedData(
-    public val groupBy: GroupBy<*, *>
+    public val dataFrame: DataFrame<*>,
+     public val keys: List<String>
 ) : TableData {
-    public val keys: List<String> = groupBy.keys.columnNames()
-    public val origin: NamedData = NamedData(groupBy.concat())
-
+    //public val keys: List<String> = groupBy.keys.columnNames()
```
Comment (devcrocod) [inline kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/ir/data/GroupedData.kt:25]: remove

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt [hunk 6]
```diff
@@ -5,14 +5,15 @@
...
+import org.jetbrains.kotlinx.dataframe.DataFrame
 import org.jetbrains.kotlinx.kandy.ir.data.GroupedData
 import org.jetbrains.kotlinx.kandy.ir.data.NamedData
 import org.jetbrains.kotlinx.kandy.ir.data.TableData
 import org.jetbrains.kotlinx.kandy.letsplot.internal.MERGED_GROUPS
 import kotlin.reflect.typeOf
 
 internal fun GroupedData.mergedKeys(): List<String> = buildList {
-    val dataFrame = origin.dataFrame
+    val dataFrame = dataFrame
```
Comment (devcrocod) [inline kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt:16]: For what? you are not overriding the dataframe

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt [hunk 7]
```diff
@@ -47,8 +48,8 @@ internal fun process(data: NamedData): Map<String, List<*>> {
 
 internal fun TableData.wrap(): Map<String, List<*>> {
     return (when (this) {
-        is NamedData -> process(this)
-        is GroupedData -> process(origin) + (MERGED_GROUPS to mergedKeys())
+        is NamedData -> process(dataFrame)
+        is GroupedData -> process(dataFrame) + (MERGED_GROUPS to mergedKeys())
         else -> TODO()
```
Comment (devcrocod) [inline kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt:53]: Exception

Comment (devcrocod) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt:53]: and is an else branch required here?

**Review by devcrocod — CHANGES_REQUESTED** [review]

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt (hunk 7 above)
Comment (AndreiKingsley) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt:53]: yes, for possible TableData implementations

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt (hunk 6 above)
Comment (AndreiKingsley) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt:16]: What "for what"? GroupData properties was changed

File: kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/contexts.kt [hunk 8]
```diff
@@ -473,11 +479,23 @@ public abstract class LayerContext(parent: LayerCollectorContext) : LayerContext
...
  * @property _layers List of layers derived from the grouped dataset. These layers are managed within the broader [LayerPlotContext].
  */
-public class GroupedContext(
-    override val _datasetIndex: Int,
-    override val _plotContext: LayerPlotContext
+public abstract class WithGroupedDataLayerCollectorContext<T, G>(
+    internal val groupBy: GroupBy<T, G>,
 ) : LayerCollectorContext() {
-    override val _layers: MutableList<Layer> = _plotContext.layers
+    abstract override val _datasetIndex: Int
```
[the remark this answers is not in the document]
Comment (devcrocod) [reply kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/contexts.kt:485]: Why need override _datasetIndex here?

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt (hunk 6 above)
Comment (devcrocod) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt:16]: ye, but for this you don’t need to take the dataframe into a separate variable

File: kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt (hunk 7 above)
Comment (devcrocod) [reply kandy-lets-plot/src/main/kotlin/org/jetbrains/kotlinx/kandy/letsplot/translator/data.kt]: okay, so need Exception for else

**Review by devcrocod — APPROVED** [review]

**Pull request merged**